jenkins/pipelines/ci/tiflash/tiflash-build-common.groovy

```diff
@@ -0,0 +1,891 @@
+            -DRUN_HAVE_STD_REGEX=0 \\
+            -G '${generator}'
+        python3 ${fix_compile_commands} ${include_flag} \\
+            --file_path=compile_commands.json \\
+            --load_diff_files_from "/tmp/tiflash-diff-files.json"
+        python3 ${run_clang_tidy} -p \$(realpath .) -j \$NPROC --files ".*/tiflash/dbms/*"
+        """
+    }
+}
+
+def buildStage(repo_path, build_dir, install_dir, proxy_cache_ready) {
+    stage('Build Dependency and Utils') {
+        parallel (
+            "Cluster Manage" : {
+                buildClusterManage(repo_path, install_dir)
+            },
+            "TiFlash Proxy" : {
+                buildTiFlashProxy(repo_path, proxy_cache_ready)
+            }
+        )
+    }
+    stage('Configure Project') {
+        cmakeConfigureTiFlash(repo_path, build_dir, install_dir, proxy_cache_ready)
+    }
+    stage('Build TiFlash') {
+        parallel (
+            "Format Check" : {
+                clangFormat(repo_path)
+            },
+            "Build TiFlash" : {
+                buildTiFlash(repo_path, build_dir, install_dir)
+            }
+        )
+    }
+}
+
+def uploadCCache(repo_path, target_dir) {
+    if (params.ENABLE_CCACHE && params.UPDATE_CCACHE) {
+        def ccache_tag = getBuildSuffix(repo_path)
+        def ccache_source = "/home/jenkins/agent/ccache/${ccache_tag}.tar"
+        dir(target_dir) {
+            sh"""
+            rm -rf ccache.tar
+            tar -cf ccache.tar .ccache
+            cp ccache.tar ${ccache_source}
+            """
+        }
+    } else {
+        echo "skip because ccache refresh is disabled"
+    }
+}
+
+def uploadProxyCache(repo_path, install_dir) {
+    if (!params.BUILD_TIFLASH) {
+        echo "skip because proxy is not built or packaged"
+    } else if (params.UPDATE_PROXY_CACHE) {
+        def proxy_suffix = getProxySuffix(repo_path)
+        def proxy_commit_hash = null
+        def status = true;
+        dir("${repo_path}/contrib/tiflash-proxy") {
+            proxy_commit_hash = sh(returnStdout: true, script: 'git log -1 --format="%H"').trim()
+        }
+        def cache_source = "/home/jenkins/agent/proxy-cache/${proxy_commit_hash}-${proxy_suffix}"
+        def suffix = getFileSuffix()
+        dir(install_dir) {
+            sh"""
+            cp libtiflash_proxy.${suffix} ${cache_source}
+            """
+        }
+    } else {
+        echo "skip because proxy cache refresh is disabled"
+    }
+}
+
+def uploadBuildArtifacts(install_dir) {
+    if (params.ARCHIVE_ARTIFACTS) {
+        def basename = sh(returnStdout: true, script: "basename '${install_dir}'").trim()
+        dir("${install_dir}/../") {
+            sh """
+            tar -czf '${basename}.tar.gz' '${basename}'
+            """
+            archiveArtifacts artifacts: "${basename}.tar.gz"
+        }
+    } else {
+        echo "skip because archiving is disabled"
+    }
+}
+
+def uploadBuildData(repo_path, build_dir) {
+    if (params.ARCHIVE_BUILD_DATA) {
```

Comment:
Shal we remove .git and contrib here as well ?

Reply:
I don't think so. TiPB and Kvptoto generates files into their directory. There won't be many build files, only 10-30 small files